drools-ansible-rulebook-integration-api/src/main/java/org/drools/ansible/rulebook/integration/api/rulesengine/RulesSetEventStructure.java

```diff
@@ -0,0 +1,308 @@
+package org.drools.ansible.rulebook.integration.api.rulesengine;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.text.similarity.LevenshteinDistance;
+import org.drools.ansible.rulebook.integration.api.domain.Rule;
+import org.drools.ansible.rulebook.integration.api.domain.RuleContainer;
+import org.drools.ansible.rulebook.integration.api.domain.RulesSet;
+import org.drools.ansible.rulebook.integration.api.domain.conditions.MapCondition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * This class holds the event paths from the rules set.
+ * Then, when the first event comes, validate the event paths comparing with incoming event structure.
+ * If the event path doesn't match with the incoming event structure, suggest the possible typo or missing node.
+ */
+public class RulesSetEventStructure {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RulesSetEventStructure.class);
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    // key: event path split into nodes
+    // value: list of rule names that use the event path
+    private Map<List<String>, List<String>> eventPathMap = new HashMap<>();
```

Comment:
I believe this should work, nevertheless having a `List` as a key for a `Map` is admittedly a bit odd an could cause hard to find bugs in future. My suggestion is to wrap this list into an immutable object (and if possible also precompute its hashcode) and use that object as the key for this map.

Reply:
@mariofusco Thanks! I'll do that.

Reply:
Introduced `EventPath`. Thanks!